animated-drawable/src/main/java/com/facebook/fresco/animation/bitmap/preparation/DropFramePreparationStrategy.java

```diff
@@ -0,0 +1,36 @@
+/*
+ * Copyright (c) 2017-present, Facebook, Inc.
+ *
+ * This source code is licensed under the MIT license found in the
+ * LICENSE file in the root directory of this source tree.
+ */
+package com.facebook.fresco.animation.bitmap.preparation;
+
+import com.facebook.common.logging.FLog;
+import com.facebook.fresco.animation.backend.AnimationBackend;
+import com.facebook.fresco.animation.bitmap.BitmapFrameCache;
+
+/**
+ * Frame preparation strategy to prepare the next n frames
+ */
+public class DropFramePreparationStrategy
```

Comment:
This is misleading. I think what you want to do is to keep the existing preparation strategy (i.e. prepare the next N frames ahead of time) but instead of scheduling the next frame in the drawable, re-schedule the same one. I had a prototype of AnimatedDrawable2 in the beginning that had 2 modes, one to delay, one to drop. This way, you don't need to change many Fresco-specific files and can just pass the flag down to the AnimatedDrawable instead. This should simplify the code a lot.

It looked like this:
```java
  @Override
  public void draw(Canvas canvas) {
    if (mAnimationBackend == null) {
      return;
    }
    mAnimationListener.onAnimationFrame(this, mCurrentFrame);
    if (mCurrentFrame == 0 && mCurrentLoopCount != 0) {
      mAnimationListener.onAnimationRepeat(this);
    }
    // TODO: only draw if close to target frame render time, otherwise draw previous frame
    // in case something else invalidates this the animation would be too fast otherwise.
    long actualRenderTimeStart = now();
    // TODO: check if frame drawn
    boolean frameDrawn = mAnimationBackend.drawFrame(this, canvas, mCurrentFrame);
    long actualRenderTimeEnd = now();
    if (FLog.isLoggable(FLog.VERBOSE)) {
      FLog.v(TAG,
          "Animation jitter: %s ms. Render time: %s ms. Frame drawn: %s",
          actualRenderTimeStart - mTargetFrameRenderTime,
          actualRenderTimeEnd - actualRenderTimeStart,
          frameDrawn);
    }
    if (!mIsRunning) {
      return;
    }
    if (frameDrawn) {
      mTargetFrameRenderTime += mAnimationBackend.getFrameDurationMs(mCurrentFrame);
    } else if (mFrameDelayMethod == FRAME_DELAY_METHOD_DROP) {
        mTargetFrameRenderTime += mAnimationBackend.getFrameDurationMs(mCurrentFrame);
        // TODO: what if frame duration < 16ms? should we just skip it instead since 60fps should be enough?
        // Maybe add limitFPS setting
        int droppedFramesThreshold = 10;
        while (mTargetFrameRenderTime < actualRenderTimeEnd && droppedFramesThreshold > 0) {
          // we need to drop frames
          if (FLog.isLoggable(FLog.VERBOSE)) {
            FLog.v(TAG, "Dropped a frame. Count: %s", mDroppedFrames);
          }
          if (!increaseFrameCount()) {
            // TODO: onAnimationDone?
            mAnimationListener.onAnimationStop(this);
            return; // animation done
          }
          mDroppedFrames++;
          droppedFramesThreshold--;
          mTargetFrameRenderTime += mAnimationBackend.getFrameDurationMs(mCurrentFrame);
        }
      } else if (mFrameDelayMethod == FRAME_DELAY_METHOD_DELAY) {
        mTargetFrameRenderTime = now() + mAnimationBackend.getFrameDurationMs(mCurrentFrame);
      }
    scheduleNextFrameOrNotifyAnimationFinished();
  }
```